### ISSUE TITLE
Transaction is not effective

### ISSUE THREAD

**Issue by junrenge** [database-rider/database-rider#482]
I use transactional  like this:
@Test
@DataSet(value ="users.yml", strategy = SeedStrategy.REFRESH, transactional = true)
     public void xxxxxx(){

      }
but data is still in database after I run the test.

**Comment by rmpestano** [comment]
Hi @junrenge, the transaction attribute just ensures a transaction is active so you can insert data in your tests without opening and commit the transaction, `it won't rollback` after the test execution. For example: https://github.com/database-rider/database-rider/blob/8ad04e5f3575fe5424de2bbe4f557128759e9320/rider-core/src/test/java/com/github/database/rider/core/TransactionIt.java#L30-L56

**Comment by fgonzalez-seatcode** [comment]
I'm using Spring TX autorollback provided by `@DataJpaTest` and altough test code is rolledback 

`@DataSet(executeScriptsBefore = ["/db/test/scripts/setup.sql"])`

_scriptsBefore_ aren't being rolledback :(

I guess this commit can be interfering
    
https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L420

could be?

**Comment by rmpestano** [comment]
Hi @fgonzalez-seatcode, that's a feature not a bug 😅. In dbrider/dbunit the aim is to prepare the dabatase **before** each test and to ignore/delete what you have as left over in db from previous tests.  So instead of relying on rollbacks we explicitly declare how we want the db before the tests executes and also how we expect the db after tests execution via `@ExpectedDataset`. I might be wrong but if you have the data you inserted during the test rollback by Spring I think `@ExpectedDataset` won't work.

**Comment by fgonzalez-seatcode** [comment]
Hi @rmpestano!

Don't take me wrong. I didn't mean neither it's a bug nor it's by design. Actually I don't know it. 

However, in order to be compatible with legacy code every test in my codebase relies heavily on particular SQL setup file. So, I need to reuse those SQL files and therefore I can't use YAML format here, in setup data.

So my test look like this:

```
@Test
@DataSet(executeScriptsBefore = ["/db/test/scripts/setup_test_data.sql"])
@ExpectedDataSet(value = "/db/test/scripts/test_expectation.yml")
void 'should reindex some stats table'() {

    // When
    repository.reindexStatsFor("f1c8f8dc-4950-4e83-9aa4-0d3ce83a3042")

}
```
Everything is working as expected including `@ExpectedDataSet`. Everything is executed under transaction:

- Setup script driven by DBrider
- repository class actions
- 

after those three things SpringBoot rollback the TX so DB is not "polluted" with test job.

I could skip the exceution of this [DBrider code line](https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L420) using my IDE debugger and I could verify that everything is rolledback as expected. However, this commit line provokes setup work to be persisted and can't be rolledback as repository work is.

**Comment by rmpestano** [comment]
Got it, however I'm not sure how to fix it without breaking non spring tests, maybe we can check if the connection has autoCommit set to true and then do not try to commit in that. Can you share a sample project which reproduces the issue?

**Comment by fgonzalez-seatcode** [comment]
Sure, just let me some time to code it.

Meanwhile, another experiment:


Using original SQL file with the dataset 🔴 -> TX its NOT completely rolledback 
```
@DataJpaTest
@DataSet(executeScriptsBefore = ["/db/test/scripts/setup.sql"])
class RepositoryIT {
```

Converting manually this SQL file to YML 🟢  -> TX it completely rolledback
```
@DataJpaTest
@DataSet("/db/test/scripts/setup.yml")
class RepositoryIT {
```

**Comment by fgonzalez-seatcode** [comment]
It seems SQL scripts execution does some TX management that YML datasets don't, probably [here](https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L420).

Maybe `@DataSet` annotation can be enhanced with something like this: 

`@DataSet(executeScriptsBefore = ["/db/test/scripts/setup.sql"], commitSQLStatements=false)`

**Comment by rmpestano** [comment]
> It seems SQL scripts execution does some TX management that YML datasets don't, probably [here](https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L420)

Yeah @DataSet data is managed by dbunit so definitely that commit is the culprit. I think we can avoid that commit if autoCommit is set to true but I didn't have time to try that.

**Comment by rmpestano** [comment]
Hi @fgonzalez-seatcode, I'm trying to come up with an example which reproduces the issue, this is the test:

```
@DataJpaTest
@DBRider
public class SpringBootDataJpaRollbackTest {

    @Autowired
    UserRepository userRepository;


    @BeforeAll
    @DataSet("users.yml")
    static void beforeAll() {

    }

    @Test
    @DataSet(executeScriptsBefore = "/scripts/addUser.sql")
    @ExpectedDataSet(value = "expectedAllUsers.yml", orderBy = "id", ignoreCols = "id")
    public void shouldListUsers() {
        assertThat(userRepository).isNotNull();
        assertThat(userRepository.count()).isEqualTo(4);
        assertThat(userRepository.findByEmail("junit5@mail.com")).isEqualTo(new User(99));
        userRepository.save(new User("bdd@cucumber.com", "cucumber"));
    }
```

`users.yaml`:

```
USERS:
  - ID: 1
    EMAIL: "dbunit@gmail.com"
    NAME: "dbunit"
  - ID: 2
    EMAIL: "rmpestano@gmail.com"
    NAME: "rmpestano"
  - ID: 3
    EMAIL: "springboot@gmail.com"
    NAME: "springboot"


```

This is the `expectedDataSet`:

```
USERS:
  - ID: 1
    EMAIL: "dbunit@gmail.com"
    NAME: "dbunit"
  - ID: 2
    EMAIL: "rmpestano@gmail.com"
    NAME: "rmpestano"
  - ID: 3
    EMAIL: "springboot@gmail.com"
    NAME: "springboot"
  - ID: 4
    EMAIL: "junit5@mail.com"
    NAME: "junit5"
  - ID: 5
    EMAIL: "bdd@cucumber.com"
    NAME: "cucumber"
```

And the addUser.sql:

```
INSERT INTO USERS
VALUES (-99, 'junit5@mail.com', 'junit5');
```

The issue is that the user added via script is not present in expectedDataset after the test execution (as I commented [here](https://github.com/database-rider/database-rider/issues/482#issuecomment-1304454507)):

<img width="1405" alt="Screen Shot 2023-01-22 at 12 29 42" src="https://user-images.githubusercontent.com/1592273/213913429-5218619b-65af-4d9d-850a-e68fc07e17fb.png">

Does that work for you?

If I remove cucumber user from the expectedDataset then the test passes.

You can find the [code here](https://github.com/database-rider/database-rider/blob/2b7f7908a5c807951c562f8669b59e218086da1f/rider-examples/spring-boot-dbunit-sample/src/test/java/com/github/database/rider/springboot/SpringBootDataJpaRollbackTest.java#L45).

**Comment by fragonib** [comment]
Not really. It's not the same thing. I guess I'm addressing a different issue. 

I forked your project and make a PR just to show you what I mean https://github.com/fragonib/database-rider/pull/1/files

I expected both to be rolled back:

- [SpringBootDataJpaRollbackWithDataSetTest](https://github.com/fragonib/database-rider/blob/a3def6a4e50a9b0bf99dbd3ba23cf908ff753771/rider-examples/spring-boot-dbunit-sample/src/test/java/com/github/database/rider/springboot/SpringBootDataJpaRollbackWithDataSetTest.java)
- [SpringBootDataJpaRollbackWithScriptsTest](https://github.com/fragonib/database-rider/blob/b9ccc230d2586c2c5129c9f2e479da017a314d10/rider-examples/spring-boot-dbunit-sample/src/test/java/com/github/database/rider/springboot/SpringBootDataJpaRollbackWithScriptsTest.java)

 but last is not because this [DBRider line](https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L420)

**Comment by rmpestano** [comment]
Hi @fragonib, thanks for the example. I think your problem is because you are mixing rider-spring with junit5. For junit5 integration you need to use the rider extension, `DBRider` from junit5 module: https://github.com/database-rider/database-rider/blob/2b7f7908a5c807951c562f8669b59e218086da1f/rider-examples/spring-boot-dbunit-sample/src/test/java/com/github/database/rider/springboot/SpringBootDataJpaRollbackTest.java#L7

**Comment by fgonzalez-seatcode** [comment]
In my example I'm running a Spring test because of annotation `@DataJpaTest` so AFAIK I should use `dbrider-spring` integration and that is what I'm using here. So, please have a look again to my example shown on my [comment](https://github.com/database-rider/database-rider/issues/482#issuecomment-1399598332) before and you'll see two classes that are running same goal. 

First works because is rolled back correctly, however the later isn't rolled back because this [DBRider line](https://github.com/database-rider/database-rider/blob/master/rider-core/src/main/java/com/github/database/rider/core/dataset/DataSetExecutorImpl.java#L421)

**Comment by rmpestano** [comment]
> @DataJpaTest so AFAIK I should use dbrider-spring

dbrider-spring and JUnit5 will cause you issues because we have a spring integration in junit5. You should be able to use `@DataJpaTest` and `com.github.database.rider.junit5.api.DBRider; `, I have that on the example I posted above. 

I can have another look over the weekend.
